src/beamlime/constructors/factories.py

```diff
@@ -0,0 +1,185 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+
+from __future__ import annotations
+
+from contextlib import contextmanager
+from typing import Any, Callable, Dict, Iterator, Type
+
+from .inspectors import ProductType
+from .providers import (
+    Product,
+    Provider,
+    ProviderGroup,
+    ProviderNotFoundError,
+    UnknownProvider,
+)
+
+
+class Factory:
+    """
+    Dependency injection object.
+
+    Notes
+    -----
+    Currently it does not inject dependencies for magic variables,
+    i.e. *args and **kwargs.
+    """
+
+    def __init__(self, *initial_prov_grs: ProviderGroup) -> None:
+        """
+        Initializes a factory with multiple provider groups.
+        """
+        from .providers import merge
+
+        self.providers = merge(*initial_prov_grs)
+
+    @property
+    def catalogue(self) -> frozenset[ProductType]:
+        """Frozen set of the product type(s) this factory can manufacture."""
+        return frozenset(self.providers.keys())
+
+    def __iter__(self) -> Iterator[ProductType]:
+        """Return an iterator of the product type(s) this factory can manufacture."""
+        return iter(self.catalogue)
+
+    def __len__(self) -> int:
+        """Return the number of product type(s) this factory can manufacture."""
+        return len(self.catalogue)
+
+    def _call_provider(self, product_type: Type[Product]) -> Product:
+        """
+        Build keyword argument dependencies of the provider
+        and call the provider with them.
+        """
+        provider = self.providers[product_type]
+        keyword_arguments = {
+            arg_name: self[arg_spec.dependency_type]
+            for arg_name, arg_spec in provider.arg_dep_specs.items()
+            if arg_spec.dependency_type in self or not arg_spec.is_optional()
+        }
+        return provider(**keyword_arguments)
+
+    def _inject_attributes(self, product: Product, product_type) -> Product:
+        """
+        Build and inject attribute dependencies of the ``product``.
+        If ``product`` is missing any attributes that are type-hinted,
+        the attribute(dependency) will be assembled and injected.
+
+        Raises
+        ------
+        ProviderNotFoundError
+            If any type-hinted attributes are not populated in ``product`` and
+            any providers for the missing attributes are not found.
```

Comment:
Now the ``self.providers`` is ``ProviderGroup`` and it raises ``ProviderNotFoundError`` instead of ``KeyError`` from the ``__getitem__``. I'll add type-hint in the init as well.